408 add change contact type action

njogz opened the pull request:
# Description

Enable the ability to edit a contact type and add a new flag that enables the edit action to work with downloaded docs.

medic/cht-conf#408

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

dianabarsan commented on src/fn/edit-contacts.js:
```diff
@@ -9,13 +9,32 @@ const toDocs = require('./csv-to-docs');
+  let overwriteAllFiles = false;
+
+  const saveJsonDoc = doc => {
+    if(fs.exists(`${docDirectoryPath}/${doc._id}.doc.json`) && !args.updateOfflineDocs && !args.force && !overwriteAllFiles) {
```
I think the code would be much easier to follow if we simplify this condition and put the "simple" action up top: 

So something like: 
```
if (args.updateOfflineDocs || args.force || overwriteAllFiles) {
  return writeFile(doc);
}

if (!fs.exists(filepath)) {
   return writeFile(doc);
}

const userSelection = userPrompt.keyInSelect();
if (userSelection === undefined) {
  throw error; 
}

if (userSelection === 1) {
  overwriteAllFiles = true;
}
return writeFile(doc);
````

dianabarsan commented on src/fn/edit-contacts.js:
```diff
@@ -135,17 +154,36 @@ function processCsv(docType, cols, row, uuidIndex, toIncludeIndex, documentDocs)
-const processDocuments =  async (csv, ids, db, args) => {
-  const documentDocs = await fetchDocumentList(db, ids);
-  return  processDocs('contact', csv, documentDocs, args);
+const fetchDocumentList = async (db, ids, args) => {
```
I think we should pull this out into a separate helper library. 
More actions could benefit from this, including `move-contacts`.
One improvement we could add (maybe later, once the function is used in multiple places) is to add a _rev check where, even if we have a copy of the doc on disk, the server could have an updated version and `upload_docs` would later fail because of a conflict.

dianabarsan reviewed (commented):
Great work! Thanks for the quick turnaround! 
I think this is going in the right direction. I left just some stylistic comments inline. 

It would be really cool if we added some `type` validation to `csv-to-docs`, now that it's not a reserved column name. 
Things like: 
- making sure that we're only allowing changing types for contacts (so don't allow a "data_record" to turn into a "task")
- making sure that the type that we set is correct and is a configured type.
- making sure that we're not changing a person to a place and vice versa. 

These are nice-to-havs for me, and since we're pressed with time, we could create additional issues to add these as improvements.

njogz commented:
Thanks for the review @dianabarsan. I have addressed your comments and creating tickets for the improvements you identified. You are on holiday so I have added @latin-panda as a reviewer as well.

craig-landry commented:
@njogz Before we run this in production please follow up with @binokaryg to organize a test against real data so we don't end up testing in production.

njogz commented:
> @njogz Before we run this in production please follow up with @binokaryg to organize a test against real data so we don't end up testing in production.

Yes. @binokaryg has already tested it but we will have another round of testing before the Muso change.

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+    include_docs: true,
+  });
+
+  const missingDocumentErrors = documentDocs.rows.filter(row => !row.doc).map(row => `Document with id '${row.key}' could not be found.`);
```
```suggestion
  const missingDocumentErrors = documentDocs.rows
    .filter(row => !row.doc)
    .map(row => `Document with id '${row.key}' could not be found.`);
```

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+    throw Error(missingDocumentErrors);
+  }
+
+  const documentTypeErrors = documentDocs.rows.filter(row => !DOC_TYPES.includes(row.doc.type)).map(row => ` Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
```
```suggestion
  const documentTypeErrors = documentDocs.rows
    .filter(row => !DOC_TYPES.includes(row.doc.type))
    .map(row => ` Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
```

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+
+  const documentTypeErrors = documentDocs.rows.filter(row => !DOC_TYPES.includes(row.doc.type)).map(row => ` Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
+  if (documentTypeErrors && documentTypeErrors.length) {
+    throw Error(documentTypeErrors);
```
I think a way to improve performance is looping through the docs to find errors once, because `filter` will run through all the records anyways, so instead of looping 4 times (filter, map and then filter map again) we can use `forEach`. (Thinking if we need to process many docs)

```js
const errors =  { invalidDocType: [], docNotFound: [] };

documentDocs.rows.forEach(row => {
   if (!row.doc) {
      errors.docNotFound.push(`Document with id '${row.key}' could not be found.`);
      return;
   }

   if (!DOC_TYPES.includes(row.doc.type)) {
     errors.invalidDocType.push(`Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
   }
});

// then since we initialised the errors arrays, we just check of length
if (errors.docNotFound.length || errors.invalidDocType.length) {
   throw Error([...errors.docNotFound, ...errors.invalidDocType]); // errors are printed in order, all the not found and then the invalid type. Check if we can do [...var] in node versions
}
```

latin-panda replied:
And once the user fixes all error then we preform the success path, as you have it in L23 :)

njogz replied:
Good suggestion!

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+const DOC_TYPES = ['district_hospital', 'health_center', 'clinic', 'person', 'user', 'user-settings', 'contact'];
+
+const jsonDocPath = (directoryPath, docID) => `${directoryPath}/${docID}.doc.json`;
+const fetchDocumentListFromDB = async (db, ids) => {
```
```suggestion

const fetchDocumentListFromDB = async (db, ids) => {
```
space between functions

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+      }
+    });
+    if(missingDocs.length > 0) {
+      const docsFromDB = await fetchDocumentListFromDB(db, missingDocs);
```
the exceptions are bubble up and catched?

njogz replied:
Yes. I have added a test case for this.

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+        docs[id] = fs.readJson(docPath);
+      }
+    });
+    if(missingDocs.length > 0) {
```
```suggestion
    if (missingDocs.length > 0) {
```
IDE can fix spacings, maybe a setting the preferences.

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+    let docs = {};
+    ids.forEach(id => {
+      const docPath = jsonDocPath(args.docDirectoryPath, id);
+      if(!fs.exists(docPath)) {
```
```suggestion
      if (!fs.exists(docPath)) {
```

latin-panda commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,47 @@
+
+module.exports = async (db, ids, args) => {
+  if(args.updateOfflineDocs) {
+    info('Loading offline doc(s)');
```
I think this function is hard to read, the missing docs job can be a function and return the ids to fetch later/
do we need to check for missing docs only for offline? do we need do it in online too?

njogz replied:
I have added a doc string to better explain what it does. We only fetch docs from the DB only if we don't find them in the directory. Let me know if I should still refactor.

latin-panda commented on test/fn/edit-contacts.spec.js:
```diff
@@ -192,6 +194,139 @@ describe('edit-contacts', function() {
+
+  it('should load documents from provided directory if relevant argument is passed', async () => {
+    const db = sinon.stub(pouch, 'allDocs');
+    sinon.stub(
```
better to break in every function when having more than one. Breaking in parameter is good when having many or too long.
Same with the other tests below
```js
sinon
  .stub(environment, 'extraArgs')
  .get(() => ['--columns=type', '--files=contact.type.csv', '--updateOfflineDocs=true', `--docDirectoryPath=${editedJsonDocs}`]);
```

latin-panda reviewed (changes requested):
Nice job!! I added few comments in line, nothing big.

Pls reference the new issues to this one so when ppl is picking the tickets they can see the full context of the improvement :)

njogz commented on src/fn/edit-contacts.js:
```diff
@@ -175,11 +175,15 @@ const parseExtraArgs = (projectDir, extraArgs = []) => {
     .split(',')
     .filter(id => id);
+  
+  const updateOfflineDocs = String(args.updateOfflineDocs).toLowerCase() === 'true'? true : false;
```
Yes. Casting to a string takes care of that. I have added a couple of test cases for that though.

newtewt replied:
minimist will handle the boolean check for you.  Instead of doing a match on the string. Just return `args.updateOfflineDocs`. Then the user only has to pass `--updateOfflineDocs` or omit  for false. This means you can remove your tests for checking different invalid strings being passed for the boolean check.  See the `--accept-self-signed-certs` as a bool check. 


```
  return {
    colNames,
    csvFiles,
    docDirectoryPath: path.resolve(projectDir, args.docDirectoryPath || 'json_docs'),
    updateOfflineDocs = args.updateOfflineDocs,
    force: !!args.force,
  };
```

craig-landry commented:
@njogz what date are you and @binokaryg going to be doing the testing with real data?

njogz commented:
> @njogz what date are you and @binokaryg going to be doing the testing with real data?

An initial test is planned for today if @binokaryg gets time. If not we will do it tomorrow.

newtewt commented on test/fn/edit-contacts.spec.js:
```diff
@@ -192,6 +194,191 @@ describe('edit-contacts', function() {
+      .returns(1);
+    sinon
+      .stub(environment, 'extraArgs')
+      .get(() => ['--columns=type', '--files=contact.type.csv', `--docDirectoryPath=${editedJsonDocs}`, '--force=true',]);
```
When force is passed here. It's being parsed as part of `extraArgs`. Is the context of force different than force passed in the main args? Meaning it will and perform the actions without user prompt? If so I think it would be worth updating this to use the main args force instead of having two --force params.

binokaryg commented:
@njuguna and I just tested it with 10894 docs (10478 contacts) in a dev clone. It worked great without any hiccups.

craig-landry commented on README.md:
```diff
@@ -155,12 +156,16 @@ file(s) | Comma delimited list of files you wish to process using edit-contacts.
 | documentID3            | true             |
 
-1. Use the following command to download and edit the documents:
+2. Use the following command to download and edit the documents:
```
The syntax for ordered/numbered lists in markdown should be as this was, where the renderer will show them as 1,2,3 automatically so you don't have to go renumber them if you insert something in the middle later . That said, clearly it's not working:
![image](https://user-images.githubusercontent.com/1485206/132764268-0fd7dff7-36c4-4915-8c9f-e9c2189652ba.png)

given that, I don't know what formatting to change here to make it function as desired. Had you tried some stuff here already?

njogz replied:
I actually hadn't looked at alternatives, but your comment prompted me to. Indenting the content below the list items fixes the issue.

craig-landry replied:
Nice, thanks for taking care of that!

craig-landry commented on src/fn/edit-contacts.js:
```diff
@@ -34,16 +52,29 @@ const execute = () => {
 
     .then(() => model.exclusions.forEach(toDocs.removeExcludedField))
-    .then(() => Promise.all(Object.values(model.docs).map(saveJsonDoc)));
+    .then(() => Promise.all(Object.values(model.docs).map(doc => {
+      const saveDoc = saveJsonDoc(doc, args, overwriteAllFiles);
+
+      if(saveDoc === DOC_EXISTS) {
+        const userSelection = userPrompt.keyInSelect();
+
+        if (userSelection === undefined || userSelection === 2) {
+          throw new Error('User canceled the action.');
+        }
+        overwriteAllFiles = (userSelection === 1);
+
+        saveJsonDoc(doc, args, overwriteAllFiles);
+      }
+    })));
```
The `Promise.all` here in the original code received promises for each `saveJsonDoc` call. In the change here though, the `return` is absent, so the `map` function is just going to return an array of `undefined` and we'll have an async bug. Adding a `return` on like 75 will cover some cases, but since that's in a branch based on 67, the "else" branch needs a return as well.

I propose a change that will cover that async behavior and also make this a bit more readable. The `if(saveDoc === DOC_EXISTS)` check can be moved to a `filter` as part of this chain. That will help to make this new function strictly about what it does do, rather than having to include a filter for what it doesn't do. Once that's been done, the function can more easily be pulled out into a named function (similar to how `saveJsonDoc` is) and then called in the chain here.

The big benefit with that is a reader can just scan over this area of the code and see the chain and nice names to get the gist of what's going on, without having to dive into all the details to reason about it.

njogz replied:
Very good catch! I have factored out overwrite file check from the `saveJsonDoc` function. It works better than having a filter because we might need to prompt user input on each doc.

craig-landry commented on src/fn/edit-contacts.js:
```diff
@@ -34,16 +52,29 @@ const execute = () => {
+        if (userSelection === undefined || userSelection === 2) {
+          throw new Error('User canceled the action.');
+        }
+        overwriteAllFiles = (userSelection === 1);
```
Mutating a variable that is used with different conditions here and above, especially with async behavior in the mix, is pretty fragile, running the risk of bugs now or in the future.

Based on the logic, I get the sense there's a case above on line 65 where we want to use the default (ignoring user selection) and then this case here where we want to respect the user selection. With that, line 65 can just use a named constant (set to false) and then here on this line, a new const can be created to name the user selection; something like `const userForceOverwrite = (userSelection === 1);`

njogz replied:
One of the options a user gets is to overwrite all files which is a selection we need to keep track of and not prompt on subsequent files that are to be overwritten thus the current implementation.

craig-landry commented on test/data/edit-contacts/csv/contact.fail.fetch.csv:
```diff
@@ -0,0 +1,5 @@
+09efb53f-9cd8-524c-9dfd-f62c242f1817,false,true
+0ebca32d-c1b7-5522-94a3-97dd8b3df146,true,false
+7ac33d1f-10d8-5198-b39d-9d61595292f6,FALSE,false
+4b8fc147-8e7b-4ca3-a116-0cd93a6b424d,FALSE,false
```
Please add the missing newline to this file.

craig-landry commented on test/data/edit-contacts/csv/contact.type.csv:
```diff
@@ -0,0 +1,3 @@
+documentID,type
+09efb53f-9cd8-524c-9dfd-f62c242f1817,health_center
+7ac33d1f-10d8-5198-b39d-9d61595292f6,person
```
Please add the missing newline to this file.

craig-landry commented on src/fn/edit-contacts.js:
```diff
@@ -34,16 +52,29 @@ const execute = () => {
+      if(saveDoc === DOC_EXISTS) {
+        const userSelection = userPrompt.keyInSelect();
+
+        if (userSelection === undefined || userSelection === 2) {
```
Watch out for the [magic number code smell](https://maximilianocontieri.com/code-smell-02-constants-and-magic-numbers) here.

For here and the other similar usages, please replace with a named constant so future readers can think about it logically, without needing to figure out what a "1" might mean.

craig-landry commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,63 @@
+/* eslint-disable no-console */
+const { info } = require('../lib/log');
+const fs = require('../lib/sync-fs');
+const DOC_TYPES = ['district_hospital', 'health_center', 'clinic', 'person', 'user', 'user-settings', 'contact'];
```
Minor readability request on this: Please rename this to indicate the usage of it. Yes, they are "doc types" but what is this list for?  You can bake that into the name so others can easily see what the list is all about, and if they need to change it, the name can hint as to if their change "feels right" or not.

craig-landry commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,63 @@
+    include_docs: true,
+  });
+
+  const errors =  { invalidDocType: [], docNotFound: [] };
+
+  documentDocs.rows.forEach(row => {
+    if (!row.doc) {
+        errors.docNotFound.push(`Document with id '${row.key}' could not be found.`);
+        return;
+    }
+
+    if (!DOC_TYPES.includes(row.doc.type)) {
+      errors.invalidDocType.push(`Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
+    }
+  });
```
Please move this to its own function. it will make the overall function a lot easier to read. As it is now, this function has 3 parts: load data, check for bad data, and do the normal case (transforming the data). The bulk of the lines though are taken up by error checking logic. Moving that to a separate function will allow readers to stay in that easy mode of reading as enumerated above.

I'd recommend a new function something like `findErrors`, which just does this same stuff and returns back the object like on line 15.

craig-landry commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,63 @@
+    }
+  });
+
+  if (errors.docNotFound.length || errors.invalidDocType.length) {
```
There's an unnamed concept in this code, which is: "If there are errors, stop". If we wanted to get all OOP on this, we'd have a method on that errors object to indicate emptiness, but more appropriate for JS is to use [_.isEmpty](https://docs-lodash.com/v4/is-empty/). As the code is currently though, that won't work, since the object is created with 2 properties. Those aren't strictly necessary though, and only seem to be to aid for the collection/push technique. At the moment that side effect is harmless, but if we're to make the `isEmpty` change (please do) it starts to matter and we're better served by having the object truly be empty. That might start to feel fragile since if a future coder makes the "mistake" of putting properties into it, they'll break the `isEmpty` check. That's just a sales pitch for unit testing though :) ...that brings up another good reason to move the error collecting logic out of here: it allows you to unit test error collection logic in isolation and build a stable interface around that function so other logic (like my propsed `isEmpty` check here) can safely rely on it.

craig-landry reviewed (changes requested):
@njogz I originally added some review notes from a quick pass, but noticed a possible async issue when looking again today. When digging in I also saw some opportunities for refactoring for improved readability and testability. The async issue should be addressed, but the others are not strictly required in this PR/release if we need this to hit a deadline.

craig-landry commented on src/fn/edit-contacts.js:
```diff
@@ -7,18 +7,37 @@ const pouch = require('../lib/db');
+
+// check to see if we should write/overwrite the file
+const overwriteFileCheck = (doc, args, overwriteAllFiles) => {
+  if (args.updateOfflineDocs || args.force || overwriteAllFiles || !fs.exists(jsonDocPath(args.docDirectoryPath, doc._id))) {
+    return true;
+  }
+  return false;
```
This is just a big boolean check, so you can return directly:

```suggestion
return args.updateOfflineDocs || args.force || overwriteAllFiles || !fs.exists(jsonDocPath(args.docDirectoryPath, doc._id))
```

craig-landry reviewed (approved):
Small code optimization, but approving now since that's a little thing.

dianabarsan commented on README.md:
```diff
@@ -144,22 +144,27 @@ Parameter | Description | Required
 column(s) | Comma delimited list of columns you wish to add/edit. If this is not specified all columns will be added. | No
 docDirectoryPath | This action outputs files to local disk at this destination | No. Default `json-docs`
 file(s) | Comma delimited list of files you wish to process using edit-contacts. By default, contact.csv is searched for in the current directory and processed. | No.
+updateOfflineDocs | If passed this updates the docs already in the docDirectoryPath instead of downloading form the server. | No.
```
```suggestion
updateOfflineDocs | If passed, this updates the docs already in the docDirectoryPath instead of downloading from the server. | No.
```

dianabarsan commented on src/fn/edit-contacts.js:
```diff
@@ -7,18 +7,34 @@ const pouch = require('../lib/db');
+};
+
+const saveJsonDoc = (doc, args) => {
+  const writeFile = (writeDoc) => fs.write(jsonDocPath(args.docDirectoryPath, writeDoc._id), safeStringify(writeDoc) + '\n');
+  
+  return writeFile(doc);
```
It seems strange to me that we create this `writeFile` function and then never use it, except to call it with the parameters we received. 
```suggestion
return fs.write(jsonDocPath(args.docDirectoryPath, doc._id), safeStringify(doc) + '\n');
```

dianabarsan commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,69 @@
+/* eslint-disable no-console */
```
Is this still needed?

dianabarsan commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,69 @@
+/* eslint-disable no-console */
+const { info } = require('../lib/log');
+const fs = require('../lib/sync-fs');
+const INVALID_DOC_TYPES = ['district_hospital', 'health_center', 'clinic', 'person', 'user', 'user-settings', 'contact'];
```
Aren't these "VALID_DOC_TYPES" ? I looks like we're throwing errors for all docs that don't have a type in this list, so these sound to me like they are valid doc types, instead.

dianabarsan commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,69 @@
+  ids.forEach(id => {
+    const docPath = jsonDocPath(args.docDirectoryPath, id);
+    if (fs.exists(docPath)) {
+      docs[id] = fs.readJson(docPath);
```
We could improve this by fetching the doc from the DB anyway, and return an invalid row if the _rev field is different. If the _rev is different on the server, `uploadDocs` will fail because of the conflict, right?  

We could do this as an improvement later, though.

njogz replied:
https://github.com/medic/cht-conf/issues/427

dianabarsan commented on src/fn/edit-contacts.js:
```diff
@@ -34,16 +50,31 @@ const execute = () => {
+    .then(() => Promise.all(Object.values(model.docs).map(doc => {
+      const overwriteFile = overwriteFileCheck(doc, args, overwriteAllFiles);
+
+      if(!overwriteFile) {
```
```suggestion
      if (!overwriteFile) {
```

dianabarsan commented on src/fn/edit-contacts.js:
```diff
@@ -34,16 +50,31 @@ const execute = () => {
     .then(() => model.exclusions.forEach(toDocs.removeExcludedField))
-    .then(() => Promise.all(Object.values(model.docs).map(saveJsonDoc)));
+    .then(() => Promise.all(Object.values(model.docs).map(doc => {
+      const overwriteFile = overwriteFileCheck(doc, args, overwriteAllFiles);
```
I think this deserves to be pulled out into its own separate function, and slim down `execute` size.

dianabarsan commented on src/lib/fetch-document-list.js:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  const missingDocs = [];
+  let docs = {};
```
Please always prefer `const` over `let`. 

```suggestion
  const docs = {};
```

dianabarsan reviewed (changes requested):
Nice work! I just have a couple of small comments inline.

njogz commented:
@dianabarsan I have addressed your comments. Please have another look. Thanks.

dianabarsan commented on src/fn/edit-contacts.js:
```diff
@@ -21,9 +21,26 @@ const overwriteFileCheck = (doc, args, overwriteAllFiles) => {
+  return fs.write(jsonDocPath(args.docDirectoryPath, doc._id), safeStringify(doc) + '\n');
+};
+
+const writeDocs = (docs, args, overwriteAllFiles) => {
```
I think `overwriteAllFiles` should just be initialized here, and not passed as a parameter. Is it used outside the scope of this function?

njogz replied:
You're right. No need to pass it as a parameter.

dianabarsan reviewed (approved):
⭐ ⭐ ⭐

newtewt reviewed (changes requested):
I added some suggestions about changing to using the force flag that's already available. 

Also it looks like we support `--column` or `--columns` for extra args. However all the tests are written for `--columns`. Can you add at least one test to ensure either name works?

newtewt reviewed (approved):
👍
